src/bp/lang-server/service/index.ts

```diff
@@ -45,75 +54,172 @@ export default class LanguageService {
     }
 
-    this._models[lang] = {
+    const fastTextModel = {
       name: lang,
-      path: modelPath,
+      path: fastTextModelPath,
       sizeInMb: 0,
       loaded: false
     }
+
+    const bpeModel = {
+      name: lang,
+      path: bpeModelPath,
+      sizeInMb: 0,
+      loaded: false
+    }
+    this._models[lang] = { fastTextModel, bpeModel }
   }
 
   @WrapErrorsWith(args => `Couldn't load language model "${args[0]}"`)
-  private async _loadModel(lang: string) {
+  private async _loadModels(lang: string) {
+    const fastTextModel = await this._loadFastTextModel(lang)
+    const bpeModel = await this._loadBPEModel(lang)
+    this._models[lang] = { fastTextModel, bpeModel }
+  }
+
+  private async _loadFastTextModel(lang: string): Promise<LoadedFastTextModel> {
     const usedBefore = process.memoryUsage().rss / 1024 / 1024
     const dtBefore = Date.now()
 
     const model = new toolkit.FastText.Model(false, true, true)
-    await model.loadFromFile(this._models[lang].path)
+    await model.loadFromFile(this._models[lang].fastTextModel.path)
 
     const dtAfter = Date.now()
     const usedAfter = process.memoryUsage().rss / 1024 / 1024
     const usedDelta = Math.round(usedAfter - usedBefore)
     const dtDelta = dtAfter - dtBefore
 
-    this._models[lang] = <LoadedModel>{ ...this._models[lang], model, sizeInMb: usedDelta, loaded: true }
+    const loadedModel = <LoadedFastTextModel>{
+      ...this._models[lang].fastTextModel,
+      model,
+      sizeInMb: usedDelta,
+      loaded: true
+    }
     console.log(`language model '${lang}' took about ${usedDelta}mb of RAM and ${dtDelta}ms to load`)
+
+    return loadedModel
   }
 
-  async vectorize(tokens: string[], lang: string): Promise<number[][]> {
-    const _model = this._models[lang] as LoadedModel
-    if (!_model || !_model.loaded) {
+  private async _loadBPEModel(lang: string): Promise<LoadedBPEModel> {
+    const usedBefore = process.memoryUsage().rss / 1024 / 1024
+    const dtBefore = Date.now()
+
+    const tokenizer = toolkit.SentencePiece.createProcessor()
+    tokenizer.loadModel(this._models[lang].bpeModel.path)
+
+    const dtAfter = Date.now()
+    const usedAfter = process.memoryUsage().rss / 1024 / 1024
+    const usedDelta = Math.round(usedAfter - usedBefore)
+    const dtDelta = dtAfter - dtBefore
+
+    const loadedModel = <LoadedBPEModel>{
+      ...this._models[lang].bpeModel,
+      tokenizer,
+      sizeInMb: usedDelta,
+      loaded: true
+    }
+    console.log(`bpe tonization model '${lang}' took about ${usedDelta}mb of RAM and ${dtDelta}ms to load`)
+
+    return loadedModel
+  }
+
+  async vectorize(input: string, lang: string): Promise<[number[][], string[]]> {
+    const { fastTextModel, bpeModel } = this._models[lang] as ModelSet
+    if (!fastTextModel || !fastTextModel.loaded || !bpeModel || !bpeModel.loaded) {
+      throw new Error(`One of FastText model or Sentencepiece bpe model for lang '${lang}' is not loaded in memory`)
```

Comment:
Which one ? I would split the models verification, could be nice to know if it's fasttext or bpe missing. What means BPE, btw ?

Reply:
It stands for byte pair encoding. It's actually a model used for tokenization by Sentencepiece.